In [1]:
import pandas as pd

df = pd.read_csv('amazon_products_sales_data_uncleaned.csv')
df.shape

(42675, 16)

In [2]:
df.head()

,title,rating,number_of_reviews,bought_in_last_month,current/discounted_price,price_on_variant,listed_price,is_best_seller,is_sponsored,is_couponed,buy_box_availability,delivery_details,sustainability_badges,image_url,product_url,collected_at
0,BOYA BOYALINK 2 Wireless Lavalier Microphone f...,4.6 out of 5 stars,375,300+ bought in past month,89.68,basic variant price: 2.4GHz,$159.00,No Badge,Sponsored,Save 15% with coupon,Add to cart,"Delivery Mon, Sep 1",Carbon impact,https://m.media-amazon.com/images/I/71pAqiVEs3...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29
1,"LISEN USB C to Lightning Cable, 240W 4 in 1 Ch...",4.3 out of 5 stars,"2,457",6K+ bought in past month,9.99,basic variant price: nan,$15.99,No Badge,Sponsored,No Coupon,Add to cart,"Delivery Fri, Aug 29",NaN,https://m.media-amazon.com/images/I/61nbF6aVIP...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29
2,"DJI Mic 2 (2 TX + 1 RX + Charging Case), Wirel...",4.6 out of 5 stars,"3,044",2K+ bought in past month,314.00,basic variant price: nan,$349.00,No Badge,Sponsored,No Coupon,Add to cart,"Delivery Mon, Sep 1",NaN,https://m.media-amazon.com/images/I/61h78MEXoj...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29
3,"Apple AirPods Pro 2 Wireless Earbuds, Active N...",4.6 out of 5 stars,"35,882",10K+ bought in past month,NaN,basic variant price: $162.24,No Discount,Best Seller,Organic,No Coupon,NaN,NaN,NaN,https://m.media-amazon.com/images/I/61SUj2aKoE...,/Apple-Cancellation-Transparency-Personalized-...,2025-08-21 11:14:29
4,Apple AirTag 4 Pack. Keep Track of and find Yo...,4.8 out of 5 stars,"28,988",10K+ bought in past month,NaN,basic variant price: $72.74,No Discount,No Badge,Organic,No Coupon,NaN,NaN,NaN,https://m.media-amazon.com/images/I/61bMNCeAUA...,/Apple-MX542LL-A-AirTag-Pack/dp/B0D54JZTHY/ref...,2025-08-21 11:14:29


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 42675 entries, 0 to 42674
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   title                     42675 non-null  str  
 1   rating                    41651 non-null  str  
 2   number_of_reviews         41651 non-null  str  
 3   bought_in_last_month      39458 non-null  str  
 4   current/discounted_price  30926 non-null  str  
 5   price_on_variant          42675 non-null  str  
 6   listed_price              42675 non-null  str  
 7   is_best_seller            42675 non-null  str  
 8   is_sponsored              42675 non-null  str  
 9   is_couponed               42675 non-null  str  
 10  buy_box_availability      28022 non-null  str  
 11  delivery_details          30955 non-null  str  
 12  sustainability_badges     3408 non-null   str  
 13  image_url                 42675 non-null  str  
 14  product_url               40606 non-null  str  
 

In [4]:
df.isnull().sum()

title                           0
rating                       1024
number_of_reviews            1024
bought_in_last_month         3217
current/discounted_price    11749
price_on_variant                0
listed_price                    0
is_best_seller                  0
is_sponsored                    0
is_couponed                     0
buy_box_availability        14653
delivery_details            11720
sustainability_badges       39267
image_url                       0
product_url                  2069
collected_at                    0
dtype: int64

In [5]:
df.duplicated().sum()

np.int64(0)

In [6]:
df.describe()

,title,rating,number_of_reviews,bought_in_last_month,current/discounted_price,price_on_variant,listed_price,is_best_seller,is_sponsored,is_couponed,buy_box_availability,delivery_details,sustainability_badges,image_url,product_url,collected_at
count,42675,41651,41651,39458,30926,42675,42675,42675,42675,42675,28022,30955,3408,42675,40606,42675
unique,8808,31,4413,59,2576,3466,911,12,2,42,1,298,16,8038,40606,1559
top,"Duracell Coppertop 9V Battery, 6 Count (Pack o...",4.6 out of 5 stars,25,100+ bought in past month,29.99,basic variant price: nan,No Discount,No Badge,Organic,No Coupon,Add to cart,"Delivery Mon, Sep 1",Small Business,https://m.media-amazon.com/images/I/51Xr76m2WL...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29
freq,744,6151,626,8801,825,21031,30364,40814,35664,40727,28022,6189,1341,1461,1,33


In [7]:
for col in ['rating', 'is_best_seller', 'is_sponsored', 'is_couponed']:
    print(f"\n--- {col} ---")
    print(df[col].value_counts())


--- rating ---
rating
4.6 out of 5 stars    6151
4.4 out of 5 stars    5525
4.5 out of 5 stars    5359
4.7 out of 5 stars    4664
4.8 out of 5 stars    4230
4.3 out of 5 stars    2927
4.2 out of 5 stars    2837
4.1 out of 5 stars    1959
4.0 out of 5 stars    1465
3.9 out of 5 stars    1316
3.8 out of 5 stars    1083
5.0 out of 5 stars     995
4.9 out of 5 stars     704
3.6 out of 5 stars     666
3.7 out of 5 stars     604
3.2 out of 5 stars     363
3.5 out of 5 stars     242
3.4 out of 5 stars     216
3.0 out of 5 stars     148
2.0 out of 5 stars     143
2.7 out of 5 stars      15
1.5 out of 5 stars      15
3.3 out of 5 stars       6
2.8 out of 5 stars       4
1.0 out of 5 stars       4
2.4 out of 5 stars       3
2.9 out of 5 stars       2
3.1 out of 5 stars       2
2.3 out of 5 stars       1
2.5 out of 5 stars       1
2.6 out of 5 stars       1
Name: count, dtype: int64

--- is_best_seller ---
is_best_seller
No Badge             40814
Amazon's               744
Limited time deal    

## Data Investigation Summary

### Dataset Overview
- amazon_products_sales_data, 42675, 16, a dataset representing a product listings data, not sales records

### Missing Values
- rating                     1024  legitimate null
number_of_reviews            1024   legitimate null 
bought_in_last_month         3217    legitimate null 
current/discounted_price    11749    legitimate null
buy_box_availability        14653    legitimate null
delivery_details            11720    true error
sustainability_badges       39267    legitimate null
product_url                  2069,   true error

### Data Type Issues
- rating,bought_in_last_month, price_on_variant,listed_price,number_of_reviews,current/discounted_price,collected_at

### Column-Level Issues
- rating: stored as text sentence
- is_best_seller: contains mixed/corrupted values
- price_on_variant: contains non-price data
- current/discounted_price: null when no discount exists (legitimate)
- buy_box_availability: single-value column, low analytical use

In [8]:
df_clean = df.copy()

In [9]:
df_clean['rating'] = df_clean['rating'].str.split(" ").str[0]
df_clean['rating'] = pd.to_numeric(df_clean['rating'], errors='coerce')

In [10]:
df_clean['rating'].dtype

dtype('float64')

In [11]:
print(df_clean['rating'])

0        4.6
1        4.3
2        4.6
3        4.6
4        4.8
        ... 
42670    5.0
42671    4.2
42672    4.3
42673    4.7
42674    4.4
Name: rating, Length: 42675, dtype: float64


## Cleaning Log

### rating
- Issue: stored as text sentence e.g. "4.6 out of 5 stars"
- Fix: extracted numeric value using str.split(), converted to float64

In [12]:
df_clean['number_of_reviews'].value_counts().head(10)

number_of_reviews
25     626
1      524
8      478
30     409
17     409
76     380
93     357
6      350
9      329
125    325
Name: count, dtype: int64

In [13]:
df_clean['number_of_reviews'].dtype

<StringDtype(storage='python', na_value=nan)>

In [14]:
df_clean['number_of_reviews'].str.contains(',').sum()

np.int64(15272)

In [15]:
df_clean['number_of_reviews'] = pd.to_numeric(df_clean['number_of_reviews'], errors='coerce').astype('Int64')

In [16]:
df_clean['number_of_reviews'].dtype

Int64Dtype()

In [17]:
print(df_clean['number_of_reviews'])

0         375
1        <NA>
2        <NA>
3        <NA>
4        <NA>
         ... 
42670       1
42671      20
42672      57
42673    <NA>
42674      75
Name: number_of_reviews, Length: 42675, dtype: Int64


### number_of_reviews
- Issue: comma-formatted numbers stored as text e.g. "35,882"
- Fix: removed commas with str.replace(), converted to Int64

In [18]:
df_clean['bought_in_last_month'].value_counts().head(10)

bought_in_last_month
100+ bought in past month       8801
50+ bought in past month        5967
200+ bought in past month       5645
No featured offers available    4571
300+ bought in past month       2842
500+ bought in past month       2351
1K+ bought in past month        2084
400+ bought in past month       1436
20K+ bought in past month        772
More Buying Choices              654
Name: count, dtype: int64

In [19]:
def clean_bought(value):
    if pd.isna(value):
        return None
    
    first = value.split(" ")[0]
    first = first.replace('+', '')
    
    if 'K' in first:
        return int(first.replace('K','')) * 1000
    else:
        try:
            return int(first)
        except:
            return None

In [20]:
df_clean['bought_in_last_month'] = df_clean['bought_in_last_month'].apply(clean_bought)
df_clean['bought_in_last_month'].dtype

dtype('float64')

In [21]:
df_clean['bought_in_last_month'] = pd.to_numeric(df_clean['bought_in_last_month'], errors='coerce').astype('Int64')
df_clean['bought_in_last_month'].dtype

Int64Dtype()

### bought_in_last_month
- Issue: text format e.g. "100+ bought in past month", "1K+ bought in past month", non-purchase labels mixed in
- Fix: wrote custom function to extract numeric value, convert K to thousands, handle invalid entries, converted to Int64

In [22]:
df_clean['listed_price'].value_counts().head(10)

listed_price
No Discount    30364
$79.99           500
$29.99           406
$16.61           360
$23.26           283
$43.99           241
$23.99           232
$25.99           210
$14.29           207
$27.33           205
Name: count, dtype: int64

In [23]:
df_clean['listed_price'].dtype


<StringDtype(storage='python', na_value=nan)>

In [24]:
df_clean['listed_price'] = df_clean['listed_price'].str.replace('$', '', regex=False)
df_clean['listed_price'] = pd.to_numeric(df_clean['listed_price'], errors='coerce')
df_clean['listed_price'].dtype

dtype('float64')

### listed_price
- Issue: $ symbol stored as text, "No Discount" for full price products
- Fix: removed $ with regex=False, converted to float64, "No Discount" became NaN via errors='coerce'

In [25]:
df_clean['current/discounted_price'].value_counts().head(10)

current/discounted_price
29.99     825
119.99    695
13.99     585
24.99     502
49.99     478
59.99     474
49.95     471
25.47     461
85.99     441
34.99     377
Name: count, dtype: int64

In [26]:
df_clean['current/discounted_price'].dtype
df_clean['current/discounted_price'].str.contains('\$', regex=True).sum()

<>:2: SyntaxWarning: invalid escape sequence '\$'
<>:2: SyntaxWarning: invalid escape sequence '\$'
/tmp/ipykernel_561/1982028461.py:2: SyntaxWarning: invalid escape sequence '\$'
  df_clean['current/discounted_price'].str.contains('\$', regex=True).sum()


np.int64(0)

In [27]:
df_clean['current/discounted_price'].dtype

<StringDtype(storage='python', na_value=nan)>

In [28]:
df_clean['current/discounted_price'] = df_clean['current/discounted_price'].str.replace('$', '', regex=False)
df_clean['current/discounted_price'] = pd.to_numeric(df_clean['current/discounted_price'], errors='coerce')
df_clean['current/discounted_price'].dtype


dtype('float64')

### current/discounted_price
- Issue: stored as text, $ symbol present in some values
- Fix: removed $ with regex=False, converted to float64

In [29]:
df_clean['collected_at'] = pd.to_datetime(df_clean['collected_at'])
df_clean['collected_at'].dtype

dtype('<M8[us]')

### collected_at
- Issue: date stored as text
- Fix: converted to datetime using pd.to_datetime()

In [30]:
df_clean['price_on_variant'].value_counts().head(10)

price_on_variant
basic variant price: nan                     21031
basic variant price: 16 Count (Pack of 1)      582
basic variant price: $9.99                     324
basic variant price: $119.95                   298
basic variant price: $169.99                   264
basic variant price: $29.99                    217
basic variant price: $349.00                   211
basic variant price: $5.00                     209
basic variant price: $59.24                    205
basic variant price: $20.63                    204
Name: count, dtype: int64

In [31]:
df_clean['price_on_variant']=df_clean['price_on_variant'].str.replace("basic variant price: ",'')
df_clean['price_on_variant']= df_clean['price_on_variant'].str.replace('$', '', regex=False)
df_clean['price_on_variant']=pd.to_numeric(df_clean['price_on_variant'], errors='coerce')

In [32]:
df_clean['price_on_variant'].dtype

dtype('float64')

In [34]:
print(df_clean['price_on_variant'].value_counts().head(10))

price_on_variant
9.99      324
119.95    298
169.99    264
29.99     217
349.00    211
5.00      209
59.24     205
20.63     204
224.00    204
216.66    204
Name: count, dtype: int64


### price_on_variant
- Issue: values prefixed with "basic variant price: ", $ symbol present, non-price data mixed in e.g. "16 Count (Pack of 1)"
- Fix: removed prefix with str.replace(), removed $ with regex=False, converted to float64, invalid entries became NaN via errors='coerce'
- As you can see in the last code ran that the 'price_on_variant' has now becom eclean, usable values that are ready to be used for analysis

In [35]:
df_clean['is_best_seller'].value_counts()

is_best_seller
No Badge             40814
Amazon's               744
Limited time deal      409
Best Seller            275
Save 30%               125
Ends in                 97
Save 10%                61
Save 18%                57
Save 17%                51
Save 12%                26
Save 9%                 14
Save 77%                 2
Name: count, dtype: int64

In [36]:
valid_values = ['No Badge', 'Best Seller', "Amazon's"]
df_clean['is_best_seller'] = df_clean['is_best_seller'].where(
    df_clean['is_best_seller'].isin(valid_values), other=None
)

In [37]:
df_clean['is_best_seller'].value_counts()

is_best_seller
No Badge       40814
Amazon's         744
Best Seller      275
Name: count, dtype: int64

### is_best_seller
- Issue: promotional labels mixed in e.g. "Save 30%", "Limited time deal", "Ends in" that are not valid entries for the 'is_best_seller' column.
- Fix: defined valid values (list), replaced invalid entries with None using .where() and .isin(), .isin() runs through the code and the value thaat meet its condition is returns True. While .where() then performs a function for all true outcomes and returns None for all Fallse outcomes. This ensure only valid entries remain in the 'is_best_seller' column.

In [43]:
df_clean['rating'].mean()
df_clean['rating'] = df_clean['rating'].fillna(df_clean['rating'].mean())
df_clean.dropna(subset=['product_url'], inplace=True)

In [44]:
#now we can save the cleaned data to a new CSV file and we will run isnull() to see the new value of missing value in the cleaned data
df_clean.isnull().sum()

title                           0
rating                          0
number_of_reviews           14905
bought_in_last_month        10508
current/discounted_price    12661
price_on_variant            21823
listed_price                29748
is_best_seller                652
is_sponsored                    0
is_couponed                     0
buy_box_availability        14373
delivery_details            11440
sustainability_badges       37215
image_url                       0
product_url                     0
collected_at                    0
dtype: int64

### Missing Values
- rating: filled with column mean (honest representation of unknown ratings)
- product_url: dropped rows with missing URLs (unusable product records)
- All other nulls: left as legitimate NaN (not applicable entries). This reveals that cleaning reveals the true extent of missing data. The original null count was an undercount because bad data was hiding as text.

In [45]:
df_clean.shape

(40606, 16)

In [58]:
df_clean.head(5)

,title,rating,number_of_reviews,bought_in_last_month,current/discounted_price,price_on_variant,listed_price,is_best_seller,is_sponsored,is_couponed,buy_box_availability,delivery_details,sustainability_badges,image_url,product_url,collected_at
0,BOYA BOYALINK 2 Wireless Lavalier Microphone f...,4.6,375,300,89.68,NaN,159.00,No Badge,Sponsored,Save 15% with coupon,Add to cart,"Delivery Mon, Sep 1",Carbon impact,https://m.media-amazon.com/images/I/71pAqiVEs3...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29
1,"LISEN USB C to Lightning Cable, 240W 4 in 1 Ch...",4.3,<NA>,6000,9.99,NaN,15.99,No Badge,Sponsored,No Coupon,Add to cart,"Delivery Fri, Aug 29",NaN,https://m.media-amazon.com/images/I/61nbF6aVIP...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29
2,"DJI Mic 2 (2 TX + 1 RX + Charging Case), Wirel...",4.6,<NA>,2000,314.00,NaN,349.00,No Badge,Sponsored,No Coupon,Add to cart,"Delivery Mon, Sep 1",NaN,https://m.media-amazon.com/images/I/61h78MEXoj...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29
3,"Apple AirPods Pro 2 Wireless Earbuds, Active N...",4.6,<NA>,10000,NaN,162.24,NaN,Best Seller,Organic,No Coupon,NaN,NaN,NaN,https://m.media-amazon.com/images/I/61SUj2aKoE...,/Apple-Cancellation-Transparency-Personalized-...,2025-08-21 11:14:29
4,Apple AirTag 4 Pack. Keep Track of and find Yo...,4.8,<NA>,10000,NaN,72.74,NaN,No Badge,Organic,No Coupon,NaN,NaN,NaN,https://m.media-amazon.com/images/I/61bMNCeAUA...,/Apple-MX542LL-A-AirTag-Pack/dp/B0D54JZTHY/ref...,2025-08-21 11:14:29


### Phase 4 — Business Logic and Patterns
- With the data now cleaned here is an example of what you can now do with the data insights, that would matter to an Amazon seller or a business analyst.
1. What rating profile do best selling products have?
2. Does offering a discount increase sales?
3. Does offering a coupon increase sales?

In [ ]:

# We will group the data by the 'is_best_seller' column and calculate the mean rating for each group. The result will be rounded to 2 decimal places for better readability.
df_clean.groupby('is_best_seller')['rating'].mean().round(2)

is_best_seller
Amazon's       4.51
Best Seller    4.53
No Badge       4.38
Name: rating, dtype: float64

###
Best Seller products average 4.53 stars. No Badge products average 4.38 stars. That's a consistent gap.
Best selling products tend to have higher ratings than products with no badge. Customers are buying and endorsing higher quality products more.
If you're a seller, this tells you that, rating matters. Products with higher ratings are more likely to become best sellers.

In [61]:
#does offering a discount increase sales?
df_clean['has_discount'] = df_clean['current/discounted_price'].notna()
df_clean.groupby('has_discount')['bought_in_last_month'].mean().round(0)

has_discount
False    405.0
True     512.0
Name: bought_in_last_month, dtype: Float64

### products with a discount sell more. 512 vs 405 bought per month.

In [62]:
#For Coupons
df_clean['has_coupon'] = df_clean['is_couponed'] != 'No Coupon'
df_clean.groupby('has_coupon')['bought_in_last_month'].mean().round(0)

has_coupon
False    478.0
True     739.0
Name: bought_in_last_month, dtype: Float64

Products with coupons sell even more than products with discounts. 739 vs 478 bought per month. Coupons have a bigger impact on sales than discounts.
Here are three business findings:

Higher rated products are more likely to be best sellers
Discounted products sell more (512 vs 405)
Products with coupons sell the most (739 vs 478)

In [65]:

df_clean.head()

,title,rating,number_of_reviews,bought_in_last_month,current/discounted_price,price_on_variant,listed_price,is_best_seller,is_sponsored,is_couponed,buy_box_availability,delivery_details,sustainability_badges,image_url,product_url,collected_at,has_discount,has_coupon
0,BOYA BOYALINK 2 Wireless Lavalier Microphone f...,4.6,375,300,89.68,NaN,159.00,No Badge,Sponsored,Save 15% with coupon,Add to cart,"Delivery Mon, Sep 1",Carbon impact,https://m.media-amazon.com/images/I/71pAqiVEs3...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29,True,True
1,"LISEN USB C to Lightning Cable, 240W 4 in 1 Ch...",4.3,<NA>,6000,9.99,NaN,15.99,No Badge,Sponsored,No Coupon,Add to cart,"Delivery Fri, Aug 29",NaN,https://m.media-amazon.com/images/I/61nbF6aVIP...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29,True,False
2,"DJI Mic 2 (2 TX + 1 RX + Charging Case), Wirel...",4.6,<NA>,2000,314.00,NaN,349.00,No Badge,Sponsored,No Coupon,Add to cart,"Delivery Mon, Sep 1",NaN,https://m.media-amazon.com/images/I/61h78MEXoj...,/sspa/click?ie=UTF8&spc=MTo4NzEzNDY2NTQ5NDYxND...,2025-08-21 11:14:29,True,False
3,"Apple AirPods Pro 2 Wireless Earbuds, Active N...",4.6,<NA>,10000,NaN,162.24,NaN,Best Seller,Organic,No Coupon,NaN,NaN,NaN,https://m.media-amazon.com/images/I/61SUj2aKoE...,/Apple-Cancellation-Transparency-Personalized-...,2025-08-21 11:14:29,False,False
4,Apple AirTag 4 Pack. Keep Track of and find Yo...,4.8,<NA>,10000,NaN,72.74,NaN,No Badge,Organic,No Coupon,NaN,NaN,NaN,https://m.media-amazon.com/images/I/61bMNCeAUA...,/Apple-MX542LL-A-AirTag-Pack/dp/B0D54JZTHY/ref...,2025-08-21 11:14:29,False,False
